BedrockServer.cpp

```diff
@@ -2209,20 +2035,268 @@ void BedrockServer::_finishPeerCommand(unique_ptr<BedrockCommand>& command) {
+    SAUTOPREFIX(request);
+
+    bool fireAndForget = false;
+    if (SIEquals(request["Connection"], "forget") || (uint64_t)request.calc64("commandExecuteTime") > STimeNow()) {
+        // Respond immediately to make it clear we successfully queued it, but don't add to the socket
+        // map as we don't care about the answer.
+        SINFO("Firing and forgetting '" << request.methodLine << "'");
+        SData response("202 Successfully queued");
+        if (_shutdownState.load() != RUNNING) {
+            response["Connection"] = "close";
+        }
+        s->send(response.serialize());
+        fireAndForget = true;
+        
+        // If we're shutting down, discard this command, we won't wait for the future.
+        if (_shutdownState.load() != RUNNING) {
+            SINFO("Not queuing future command '" << request.methodLine << "' while shutting down.");
+            return nullptr;
+        }
+    } else {
+        SINFO("Waiting for '" << request.methodLine << "' to complete.");
+    }
+
+    // Get the source ip of the command.
+    char *ip = inet_ntoa(s->addr.sin_addr);
+    if (ip != "127.0.0.1"s) {
+        // We only add this if it's not localhost because existing code expects commands that come from
+        // localhost to have it blank.
+        request["_source"] = ip;
+    }
+
+    // Create a command.
+    unique_ptr<BedrockCommand> command = getCommandFromPlugins(move(request));
+    SINFO("Deserialized command " << command->request.methodLine);
+    command->socket = fireAndForget ? nullptr : s;
+
+    if (command->writeConsistency != SQLiteNode::QUORUM && _syncCommands.find(command->request.methodLine) != _syncCommands.end()) {
+        command->writeConsistency = SQLiteNode::QUORUM;
+        _lastQuorumCommandTime = STimeNow();
+        SINFO("Forcing QUORUM consistency for command " << command->request.methodLine);
+    }
+
+    // This is important! All commands passed through the entire cluster must have unique IDs, or they
+    // won't get routed properly from follower to leader and back.
+    command->id = args["-nodeName"] + "#" + to_string(_requestCount++);
+
+    // And we and keep track of the client that initiated this command, so we can respond later, except
+    // if we received connection:forget in which case we don't respond later
+    command->initiatingClientID = SIEquals(command->request["Connection"], "forget") ? -1 : s->id;
+
+    return command;
+}
+
+void BedrockServer::handleSocket(Socket* s) {
+    // Initilaize and get a unique thread ID.
+    SInitialize("socket" + to_string(_socketThreadNumber++));
+    SINFO("Socket thread starting");
+
+    // This outer loop just runs until the entire socket life cycle is done, meaning it deserializes a command,
+    // waits for it to get processed, deserializes another, etc, until the socket gets closed.
+    // This whole block is largely duplicated from `postPoll` and modified to work on a single non-blocking socket.
+    while (1) {
```

Comment:
"while" I can see that you have buttoned up the break statements to make sure that we'll exit from this loop, there's no guarantee the next engineer will do the same. We've had problems with this exact structure, i.e. nested endless while loops, getting stuck. Wouldn't it be a better practice to always include an exit condition, even if you don't necessarily expect to wait for it? (because you want to break early, for example)

Reply:
Maybe? I'm not sure exactly what you're proposing that would be more maintainable than what we've already got here.

Reply:
NAB we could instead move `exitMainLoop` up here and do `while(!exitMainLoop)`. Though I'm not sure if that'd help or hurt the readability here. Seeing `while(1)` sort of gives me pause and makes me think "oh I need to be careful".

Reply:
Yes, I think that `while(!exitMainLoop)` would be perfectly acceptable. The problem with endless while loops (again, we've experienced this _specific_ issue in the past) is that sometimes you fail to set the exit criteria in one edge case and then are stuck in the loop.

Reply:
Doing that literally changes nothing except that line. The code remains identical except that `exitMainLoop` moves up and the `while(1)` becomes `while(!exitMainLoop)`. I don't see how that's more maintainable. You can still fail to set the exit criteria in one edge case when the exit criteria is `exitMainLoop == true`.

Reply:
Yeah I agree it's still an endless loop and has all the same pitfalls but now it's just masquerading like it isn't because it's tracking some variable instead of `(1)`.